### PR TITLE
[WIP]Ports generative networks

### LINKED ISSUE #6676: Move MONAI Generative into core
As discussed in the core meeting, [MONAI Generative](https://github.com/Project-MONAI/GenerativeModels) is going to be integrated into the core MONAI repository. The plan is to do this in stages:

1. Port the main parts of the generative code (utils, networks, metrics, losses, inferers,  engines, and associated tests). This might lead to some redundancy, but we can focus on the port for now and aim to reduce redundancies further on.
2. Once these features are available in dev, we can move the [tutorials](https://github.com/Project-MONAI/GenerativeModels/tree/main/tutorials) to the [tutorial repo](https://github.com/Project-MONAI/tutorials) - this should be easy to do, mostly involving updates imports.
3. We can then move the [model zoo](https://github.com/Project-MONAI/GenerativeModels/tree/main/model-zoo) to the [model zoo repo](https://github.com/Project-MONAI/model-zoo). 

I can start working on stage 1 soon. Any thoughts/comments?

@ericspod @Warvito @wyli 

## Progress
- [x] Losses [(PR link)](https://github.com/Project-MONAI/MONAI/pull/6729)
- [x] Metrics
- [x] Utils
- [x] Networks
- [x] DIffusion schedulers
- [x] Inferers
- [x] Engines

### PR DESCRIPTION
**Pull request by marksgraham**
Partially fixes #6676  .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by marksgraham** [comment]
I'm encountering torchscript errors, which I will document as I implement each network:

- AutoEncoderKL has an an error in the `test_script_save` module due to the option to do activation checkpointing using `torch.utils.checkpoint`. I noticed other networks with activation checkpointing on MONAI, such as [SwinUNETR](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/swin_unetr.py) do not test for torchscript saving - maybe it is OK to drop this requirement for networks that use checkpointing?

**Comment by KumoLiu** [comment]
Hi @marksgraham, yes it's fine to skip the torchscript test.

File: monai/networks/nets/autoencoderkl.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+            conv_only=True,
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = nn.functional.pad(x, self.pad, mode="constant", value=0.0)
+        x = self.conv(x)
+        return x
+
+
+class ResBlock(nn.Module):
```
Comment (KumoLiu) [inline monai/networks/nets/autoencoderkl.py:126]: There is also a `ResBlock` and `AttentionBlock`.
https://github.com/Project-MONAI/MONAI/blob/782c1e616312612c9f25bf4a33986ace67f7851e/monai/networks/blocks/segresnet_block.py#L44
https://github.com/Project-MONAI/MONAI/blob/782c1e616312612c9f25bf4a33986ace67f7851e/monai/networks/nets/attentionunet.py#L90
Perhaps if it could combine, that would be better?

**Review by KumoLiu — COMMENTED** [review]
Thanks for the PR!
I leave a few comments online. 
There appear to be some duplicate block names with exist one, is it possible for us to combine some of these?

cc @ericspod @wyli @Nic-Ma

**Comment by marksgraham** [comment]
> Thanks for the PR! I leave a few comments online. There appear to be some duplicate block names with exist one, is it possible for us to combine some of these?
> 
> cc @ericspod @wyli @Nic-Ma

Thanks for reviewing - good point on the blocks, I'll see how many of them can be combined.

I've still got a few more networks to add, too, just adding bits to the PR as I go along.

**Comment by marksgraham** [comment]
> Thanks for the PR! I leave a few comments online. There appear to be some duplicate block names with exist one, is it possible for us to combine some of these?
> 
> cc @ericspod @wyli @Nic-Ma

Would it be OK to defer the merging of some of the duplicate blocks (e.g. Attention, Resblock) until porting the rest of MONAI Generative is done? Until it is fully ported, we will be supporting the [MONAI Gen repo](https://github.com/Project-MONAI/GenerativeModels) too, and keeping the code here tightly coupled to that codebase will make it much easier to push any bugfixes etc from there to here. Once the port is done, MONAI Gen will be archived and users will be redirected here, and then it will be easier to devote our efforts to more tightly integrating the ported code and MONAI Core through better used of shared blocks, etc.

**Comment by KumoLiu** [comment]
> Would it be OK to defer the merging of some of the duplicate blocks (e.g. Attention, Resblock) until porting the rest of MONAI Generative is done? 

Yes, that should be fine. We can also create another ticket for refactoring.

**Comment by ericspod** [comment]
> Would it be OK to defer the merging of some of the duplicate blocks (e.g. Attention, Resblock) until porting the rest of MONAI Generative is done?

Duplicate code/functionality isn't good and doesn't look good either. I get your reasoning for pushing off a refactor so I would suggest that we mark redundant definitions as being private (prefix _ to their names) and write in their docstrings that these shouldn't be used and will be factored our later. What we don't want is users relying on these types directly then they disappear or we replace them with shim types we have to maintain.

**Comment by marksgraham** [comment]
Ticket for the refactoring is up [here](https://github.com/Project-MONAI/MONAI/issues/7227)

File: monai/networks/nets/autoencoderkl.py [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+    Convolution-based downsampling layer.
+
+    Args:
+        spatial_dims: number of spatial dimensions (1D, 2D, 3D).
+        in_channels: number of input channels.
+    """
+
+    def __init__(self, spatial_dims: int, in_channels: int) -> None:
+        super().__init__()
+        self.pad = (0, 1) * spatial_dims
```
[the remark this answers is not in the document]
Comment (marksgraham) [reply monai/networks/nets/autoencoderkl.py:116]: This is the padding value that allows for downsampling by a factor of 2 each level; so that the output shape of the network matches the input shape (as long as the input shape is even). I can't think of a use case for changing the padding here, specifically. I suppose we might want to find a way to support a user supplying non-even inputs but I'm not even sure if MONAI's base AutoEncoder network supports that right now.

[the remark this answers is not in the document]
Comment (ericspod) [reply monai/networks/nets/autoencoderkl.py:116]: We have a [padding argument](https://github.com/Project-MONAI/MONAI/blob/c300b36b0deb38fcad5ad18825ba2007253ce007/monai/networks/nets/autoencoder.py#L109C14-L109C14) for the AutoEncoder which can be the values compatible with the Pytorch Conv classes. This will by default be set to be a value to ensure 1-stride convolutions produce outputs with the same shape as the inputs but it can be set to other things. 

The padding you have here is specifically for convolutions having an odd kernel size (eg. 3x3 or 3x3x3)? Instead of having a padding value just don't set `padding` in `Convolution` and let that class handle how to do the padding for downsampling. 

I'm not sure you need this class anyway since the downsample you want can be done with `Convolution` with `stride=2, kernel_size=3, conv_only=True`.

**Comment by KumoLiu** [comment]
Hi @marksgraham, thanks for your hard work on this PR.
But I believe the PR is too large, making it difficult to REVIEW and ensure that the unittests are all covered. Is it possible to split this PR into multiple PRs? Perhaps a network in a PR? What do you think?
cc @ericspod @Nic-Ma

**Comment by marksgraham** [comment]
Yes that seems reasonable to me! I'll split into a PR for each network over the coming week :)
